onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,244 @@
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class PadOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+
+  int GetMinSupportedOpSet(const Node& /* node */) const override {
+    // Note: before Pad-11, inputs `pads` and `constant_value` were attributes
+    return 11;
+  }
+};
+
+// Add operator related
+
+#ifdef __APPLE__
+void PadOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());  //  pads
+  model_builder.AddInitializerToSkip(node.InputDefs()[2]->Name());  //  constant_value
+  if (node.InputDefs().size() > 3) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[3]->Name());  // axes
+  }
+}
+
+Status PadOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                           const Node& node,
+                                           const logging::Logger& logger) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  auto* coreml_pad = layer->mutable_padding();
+  auto* constant_padding_type = coreml_pad->mutable_constant();  // CoreML::Specification::PaddingLayerParams_PaddingConstant
+
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  GetShape(*input_defs[0], input_shape, logger);
+  const auto input_rank = input_shape.size();
+
+  const auto& pads_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());            // pads
+  const auto& constant_value_tensor = *model_builder.GetInitializerTensors().at(input_defs[2]->Name());  // constant_value
+
+  Initializer constant_value_initializer(constant_value_tensor);
+  float constant_value = constant_value_initializer.DataAsSpan<float>()[0];
+  constant_padding_type->set_value(constant_value);
+
+  Initializer pads_initializer(pads_tensor);
+  auto pads_span = pads_initializer.DataAsSpan<int64_t>();
+
+  std::variant<InlinedVector<int64_t>, gsl::span<const int64_t>> axes_tensor_data;
+  int64_t num_axes = input_rank;
+  if (input_defs.size() > 3) {
+    // optional input axes is provided
+    const ONNX_NAMESPACE::TensorProto& axes_tensor = *model_builder.GetInitializerTensors().at(input_defs[3]->Name());
+    Initializer axes_initializer(axes_tensor);
+    auto axes_span = axes_initializer.DataAsSpan<int64_t>();
+    axes_tensor_data.emplace<1>(axes_span);
```

Comment:
thanks for the suggestions and catch up! I believe I addressed most of the comments. except here, I didn't find a way to emplace an initializer type directly into `std::variant` type object here. instead kept the second type as a gsl::span. 

I agree it would make more sense to have a `std::variant<InlinedVector<int64_t>, Initializer>` here for `axes_tensor_data`.

Reply:
using a span to refer to the data is still a good approach - it can be used with the variant.

as for setting the variant value, something like this should work:
```c++
std::variant<InlinedVector<int64_t>, Initializer> axes_tensor_data_holder;
if (input_defs.size() > 3) {
  // use initializer
  //...
  axes_tensor_data_holder.emplace<Initializer>(axes_tensor);
} else {
  // use default axes vector
  // ...
  axes_tensor_data_holder.emplace<InlinedVector<int64_t>>(input_rank);
  // ...
}

gsl::span<const int64_t> axes_tensor_data =
    std::holds_alternative<Initializer>(axes_tensor_data_holder)
        ? std::get<Initializer>(axes_tensor_data_holder).DataAsSpan<int64_t>()
        : std::get<InlinedVector<int64_t>(axes_tensor_data_holder);
```

the important thing is that we don't want a span referring to data of an object that is no longer in scope.

it may also be useful to have a helper function to create the variant and associated span.